### ISSUE TITLE
[Deprecation] Deprecate getJoinValidator API of DiscoveryPlugin and make validateJoinRequest optional

### ISSUE THREAD

**Issue by amkhar** [opensearch-project/OpenSearch#7678]
**Describe the bug**
There are 2 types of join validators :
1. Common join validators : ensureNodeCompatibility, ensureIndexCompatibility etc
2. Pluggable join validators : there are none today

Pluggable join validators are there so any discovery based plugin can run extra validator on top of common ones during validateJoinRequest flow. Today we run common join validators 3 times in node join flow. 
* First cluster manager run these validators, then sends a validateJoinRequest to data node. 
* Data node then runs the same validations with pluggable validators if defined by discovery plugin. 
* Later cluster manager again runs common validators before actually adding the node in cluster state.
Above 3 code paths are explained in https://github.com/opensearch-project/OpenSearch/issues/5100#issuecomment-1557172548

This validateJoinRequest requires cluster manager to send the full cluster state. We're already working on optimizing this code path(as part of https://github.com/opensearch-project/OpenSearch/pull/7321), but there is an opportunity to make this request optional(came up during discussions : [comment 1](https://github.com/opensearch-project/OpenSearch/issues/5100#issuecomment-1514379627) , [comment 2](https://github.com/opensearch-project/OpenSearch/issues/5100#issuecomment-1553258966) , [comment 3](https://github.com/opensearch-project/OpenSearch/issues/5100#issuecomment-1558174441) ) and let customers opt out of this check. For that we need to do 3 things :

* Assess the usage of pluggable join validators (we know ec2, azure and gce plugins are not using it)
* Deprecate the getJoinValidator API of DiscoveryPlugin interface
* Add a setting to make this request optional

**Ideal behaviour**
If no discovery plugins are using any pluggable validators, we should be able to avoid this unnecessary transport call as join validators are already running on cluster manager node twice.

**Comment by shwetathareja** [comment]
Yes, we need to get enough information on how customers are implementing this JoinValidator in their custom plugin and use case around it before we should deprecate the pluggable support in the plugin. As an alternative, I am thinking to auto opt out of this ValidateJoinRequest by dynamic checking if JoinValidators contains only buildInValidators, that means no plugin is overriding and every node during join will send that flag to Leader and accordingly leader would skip the ValidateJoinRequest altogether. This way customers/ plugins need not change anything.

**Comment by andrross** [comment]
Based on my understanding, the pluggable join validators seem like a trap. I believe the following is possible:

Assume three nodes: `leader`, `joining-node-1` and `joining-node-2`. Assume that both joining nodes implement a custom validator that is mutually incompatible, meaning neither node will claim to be compatible with a cluster that contains the other node. The following sequence can happen:

1. `leader` handles join request of `joining-node-1` at cluster state v1 -> checks pass
2. `leader` handles join request of `joining-node-2` at cluster state v1 -> checks pass
3. `leader` sends validate join request to `joining-node-1` at cluster state v1 -> checks pass
4. `leader` sends validate join request to `joining-node-2` at cluster state v1 -> checks pass
5. `leader` adds `joining-node-1` (checks pass again on leader), changing cluster state to v2
6. `leader` adds `joining-node-2` (checks pass again on leader because incompatibility is only defined in pluggable validator on joining node), changing cluster state to v3

What happens here when cluster state v3 is published? In this contrived example neither joining node will be happy that the other joining node is present in cluster state v3.

All this is a long winded way of saying that if the situation above is possible and we don't have a good way of handling it, then maybe we should remove this pluggability because it can give custom plugins a false sense of security.
